### DISCUSSION TITLE
Modbus RTU over TCP

### DISCUSSION THREAD

**Discussion by Bavarialex** [eModbus/eModbus#244] [Q&A]
Dear all,

I'm  trying to read out a converter, which gives Modbus over TCP (means TCP-protocol, but with CRC).
Is there a trick to bind RTU server to Wifi (ESP32)?

Thanks,
Alex

**Comment by Miq1** [comment]
There is no direct connection possible between Modbus RTU and TCP, for a variety of reasons (protocols, multi-client vs. single client etc.)

Therefore the eModbus bridge devices exist. These act as a server in one domain (TCP or RTU) and as a client in the other. Requests are translated, forwarded into the other domain and vice versa for the responses.

I would suggest reading the docs where the bridges are described in detail.

**Reply by Bavarialex** [reply]
Thanks for your assistance.
Sorry, but I ment RTU over TCP, Sending RTU packages via TCP/IP.
![rtu](https://user-images.githubusercontent.com/96592622/205842946-beb29003-0081-4817-9c7e-56a543af2a90.JPG)
So, using the RTU protocol, but sending over TCP.

Something like this:
ModbusServerRTU MBserver(ip, port); ?

**Reply by Miq1** [reply]
Ah, sorry, I did not get that right. So how is the transmission over IP done? Is it like the Modbus TCP protocol, but just two additional CRC bytes added or is it completely different?

**Reply by Miq1** [reply]
I googled a bit and it seems that the non-standard "Modbus RTU over TCP" means regular RTU messages are being tunneled through a TCP connection.
Provided that you are able to extract the RTU payload, you can use the CRC utility functions from ``RTUutils`` to check or add the CRC bytes proper. The docs on these functions are [here](https://emodbus.github.io/modbusclient-rtu-api#uint16_t-rtuutilscalccrc16uint8_t-data-uint16_t-len-and-uint16_t-rtuutilscalccrc16modbusmessage-msg).

**Reply by Bavarialex** [reply]
It's the other way round. Take the RTU and send it via IP.
That's what I found:
"Modbus RTU ADU is made up of: Slave address + PDU + CRC.
Modbus RTU over TCP, takes the whole standard RTU ADU (as used on serial RTU networks) and sends it over TCP (basically putting the RTU ADU inside a TCP wrapper and transferring it over a TCP interface). The receiving end must then extract the RTU ADU from the TCP layer and pass it to the serial RTU device on the serial network."

**Reply by Miq1** [reply]
Our posts have crossed - so my assumption was right and you will have to use the static RTUutils functions.

**Reply by Bavarialex** [reply]
I'm sorry, but I don't get it.
RTU includes Slave+PDU+CRC already. So this will be the message to be sent. 
No need to calculate it again...
How may this be pushed over IP to a dedicated adress?
(In Germany we say: "You don't see the forest - there are a lot of trees...") ;-)

**Reply by Miq1** [reply]
Man sieht den Wald vor Bäumen nicht? 😄 
(ich bin Deutscher, du wahrscheinlich auch und aus Bayern?)

Let us sort out the basics. You are attempting to communicate with some converter. Will it act as a client or a server? i.e., will you request something from it or will you respond to the controller's requests?

**Reply by Bavarialex** [reply]
Servus - Neuburg an der Donau... :-)

Ok, there's a solar-charger which supplies RS485. Behind that there's a converter (EPever-Wifi01) which does the conversion to IP.
This is master for the charger and server for IP.
I can reach it and get the infos via RTU over TCP (reading out with IOBroker and  SimplyModbus).
What I intend to do is getting the data via an ESP32 to display them locally.
So, using TCPclient and add ID at the beginning and CRC at the end or use the RTUmaster but not bind it to a serial but to an IP connection?
Thanks for your time - the next beer is on me...

**Reply by Miq1** [reply]
Why do you use the converter? It would be easy to connect an ESP32 to the charger via RS485 and use the ESP's WiFi to contact it. The ESP even can be run as a bridge for another Modbus TCP client to request it, but also may report to MQTT or whatever you are using.

**Reply by Bavarialex** [reply]
Sure - but the converter came with the charger and it was easy to use - with IOBroker. So I thought, this could be no big deal with esp... But I was on the woodway (german again)...

**Reply by Miq1** [reply]
The problem you are facing now is similar, I would say - you need to cope with the non-standard protocol the converter is providing. My choice would be to reduce the number of involved devices and roll my own bridge instead that will give me all the flexibility I could need. YMMV, as ever.
